// ldapts/filters/Filter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Filter {
    write(writer) {
        writer.startSequence(this.type);
        this.writeFilter(writer);
        writer.endSequence();
    }
    parse(reader) {
        return this.parseFilter(reader);
    }
    matches(_ = {}, __) {
        return true;
    }
    /**
     * RFC 2254 Escaping of filter strings
     *
     * Raw                     Escaped
     * (o=Parens (R Us))       (o=Parens \28R Us\29)
     * (cn=star*)              (cn=star\2A)
     * (filename=C:\MyFile)    (filename=C:\5cMyFile)
     */
    escape(input) {
        let escapedResult = '';
        if (Buffer.isBuffer(input)) {
            for (const inputChar of input) {
                if (inputChar < 16) {
                    escapedResult += `\\0${inputChar.toString(16)}`;
                }
                else {
                    escapedResult += `\\${inputChar.toString(16)}`;
                }
            }
        }
        else {
            for (const inputChar of input) {
                switch (inputChar) {
                    case '*':
                        escapedResult += '\\2a';
                        break;
                    case '(':
                        escapedResult += '\\28';
                        break;
                    case ')':
                        escapedResult += '\\29';
                        break;
                    case '\\':
                        escapedResult += '\\5c';
                        break;
                    case '\0':
                        escapedResult += '\\00';
                        break;
                    default:
                        escapedResult += inputChar;
                        break;
                }
            }
        }
        return escapedResult;
    }
    // tslint:disable-next-line:no-empty
    parseFilter(_) {
    }
    // tslint:disable-next-line:no-empty
    writeFilter(_) {
    }
    getObjectValue(objectToCheck = {}, key, strictAttributeCase) {
        let objectKey;
        if (typeof objectToCheck[key] !== 'undefined') {
            objectKey = key;
        }
        else if (!strictAttributeCase && key.toLowerCase() === 'objectclass') {
            for (const objectToCheckKey of Object.keys(objectToCheck)) {
                if (objectToCheckKey.toLowerCase() === key.toLowerCase()) {
                    objectKey = objectToCheckKey;
                    break;
                }
            }
        }
        if (objectKey) {
            return objectToCheck[objectKey];
        }
        return undefined;
    }
}
exports.Filter = Filter;
