// ldapts/messages/SearchEntry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const MessageResponse_1 = require("./MessageResponse");
const ProtocolOperation_1 = require("../ProtocolOperation");
const Attribute_1 = require("../Attribute");
class SearchEntry extends MessageResponse_1.MessageResponse {
    constructor(options) {
        super(options);
        this.protocolOperation = ProtocolOperation_1.ProtocolOperation.LDAP_RES_SEARCH_ENTRY;
        this.name = options.name || '';
        this.attributes = options.attributes || [];
    }
    parseMessage(reader) {
        this.name = reader.readString();
        reader.readSequence();
        const end = reader.offset + reader.length;
        while (reader.offset < end) {
            const attribute = new Attribute_1.Attribute();
            attribute.parse(reader);
            this.attributes.push(attribute);
        }
    }
    toObject(requestAttributes) {
        const result = {
            dn: this.name,
        };
        for (const attribute of this.attributes) {
            if (attribute.values && attribute.values.length) {
                if (attribute.values.length === 1) {
                    result[attribute.type] = attribute.values[0];
                }
                else {
                    result[attribute.type] = attribute.values;
                }
            }
            else {
                result[attribute.type] = [];
            }
        }
        // Fill in any missing attributes that were requested
        for (const attribute of requestAttributes) {
            if (typeof result[attribute] === 'undefined') {
                result[attribute] = [];
            }
        }
        return result;
    }
}
exports.SearchEntry = SearchEntry;
